cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/deploy/InMemoryProgramRunDispatcher.java

```diff
@@ -0,0 +1,497 @@
+   * Return the copy of the {@link ProgramOptions} including locations of plugin artifacts in it.
+   *
+   * @param options the {@link ProgramOptions} in which the locations of plugin artifacts needs to be included
+   * @param programId Id of the Program
+   * @param tempDir Temporary Directory to create the plugin artifact snapshot
+   * @param appSpec program's Application Specification
+   * @return the copy of the program options with locations of plugin artifacts included in them
+   */
+  private ProgramOptions createPluginSnapshot(ProgramOptions options, ProgramId programId, File tempDir,
+    @Nullable ApplicationSpecification appSpec, boolean isDistributed) throws Exception {
+    // appSpec is null in an unit test
+    if (appSpec == null) {
+      return options;
+    }
+
+    Set<String> files = Sets.newHashSet();
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+    builder.putAll(options.getArguments().asMap());
+    for (Map.Entry<String, Plugin> pluginEntry : appSpec.getPlugins().entrySet()) {
+      Plugin plugin = pluginEntry.getValue();
+      File destFile = new File(tempDir, Artifacts.getFileName(plugin.getArtifactId()));
+      // Skip if the file has already been copied.
+      if (!files.add(destFile.getName())) {
+        continue;
+      }
+
+      try {
+        ArtifactId artifactId = Artifacts.toProtoArtifactId(programId.getNamespaceId(), plugin.getArtifactId());
+        ArtifactDetail artifactDetail = getArtifactDetail(artifactId);
+        String peer = options.getArguments().getOption(ProgramOptionConstants.PEER_NAME);
+        copyArtifact(artifactId, artifactDetail, destFile, isDistributed, peer != null);
+      } catch (ArtifactNotFoundException e) {
+        throw new IllegalArgumentException(String.format("Artifact %s could not be found", plugin.getArtifactId()), e);
+      }
+    }
+    LOG.debug("Plugin artifacts of {} copied to {}", programId, tempDir.getAbsolutePath());
+    builder.put(ProgramOptionConstants.PLUGIN_DIR, tempDir.getAbsolutePath());
+    return new SimpleProgramOptions(options.getProgramId(), new BasicArguments(builder.build()),
+      options.getUserArguments(), options.isDebug());
+  }
+
+  /**
+   * Copies the artifact jar to the given target file. Copies artifact from ArtifactCacheService if the program run is
+   * for a tethered peer, else copies the artifact from the filesystem.
+   *
+   * @param artifactId artifact id of the artifact to be copied
+   * @param artifactDetail detail information of the artifact to be copied
+   * @param targetFile target file to copy to
+   * @throws IOException if the copying failed
+   */
+  private void copyArtifact(ArtifactId artifactId, ArtifactDetail artifactDetail, File targetFile,
+    boolean isDistributed, boolean isTetheredPeer) throws Exception {
+    if (isTetheredPeer) {
+      try (InputStream in = artifactRepository.newInputStream(Id.Artifact.fromEntityId(artifactId))) {
+        if (isDistributed) {
```

Comment:
there is a lot of nesting and duplication here, better to break it up into multiple methods.

Reply:
Changed